### PR TITLE
refactor CreateMenu to have dynamic items

### PR DESCRIPTION
**Pull request by Maftalion**
### Details
- updated `CreateMenu` to take a dynamic list of items based on `MENU_ITEM_KEYS` enum
- updated `ReportActionCompose` to open the original attachment modal when `AttachmentPicker` is selected

### Fixed Issues
<Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing>
Fixes https://github.com/Expensify/Expensify.cash/issues/1508

### Tests

1. log in and click on the + fab to verify list has new items
2. enter a convo with more than one person to see `SplitBill|AddAttachment` options when clicking the attach icon
3. enter a convo with one person to see `RequestMoney |AddAttachment` options when clicking the attach icon

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Video Walkthrough

https://user-images.githubusercontent.com/12075600/108810552-3ca9c000-7560-11eb-9ed2-ce4325973ec9.mov

#### Web

Screen Shot 2021-02-22 at 10 22 47 PM

#### Mobile Web

Screen Shot 2021-02-22 at 10 50 49 PM

#### Desktop

Screen Shot 2021-02-22 at 10 12 33 PM

#### iOS

Screen Shot 2021-02-22 at 10 07 21 PM

#### Android

Screen Shot 2021-02-22 at 10 29 25 PM

### CODE REVIEW

**Comment by Maftalion** [comment]
I have read the CLA Document and I hereby sign the CLA

File: src/pages/home/report/ReportActionCompose.js
Comment (Maftalion) [inline src/pages/home/report/ReportActionCompose.js:193]: left this icon as a `Paperclip`, might want to update to make it seem more like a menu icon

File: src/pages/home/report/ReportActionCompose.js
Comment (Maftalion) [inline src/pages/home/report/ReportActionCompose.js:199]: the attachment modal wasn't opening unless i added this. feels a bit weird so if there's a better way, please let me know.

Comment (thienlnam) [reply src/pages/home/report/ReportActionCompose.js:199]: Yeah this is a bit of a hack, I'll look into it and try to get back to you by tommorow

**Review by thienlnam — DISMISSED** [review]
Thanks for the PR! Code looks mostly good, just a couple of comments

**Comment by thienlnam** [comment]
Also, could you update your PR to fix the merge conflict and the description to update the fixed issued link, and I think you mean

> enter a convo with more one person to see RequestMoney |AddAttachment options when clicking the attach icon

Enter a convo with just one person?

File: src/components/CreateMenu.js
Comment (Julesssss) [inline src/components/CreateMenu.js:16]: Is it possible to just import `CONST.MENU_ITEM_KEYS` instead of the whole file with React Native?

File: src/components/CreateMenu.js
Comment (Julesssss) [inline src/components/CreateMenu.js:76]: It would be good to keep `MENU_ITEMS` out of the class definition, but I'm guessing we can't do that due to the 'this' reference in `this.props.onAttachmentPickerSelected`, right?

File: src/pages/home/report/ReportActionCompose.js
Comment (Julesssss) [inline src/pages/home/report/ReportActionCompose.js:197]: Could we instead use `this.state.isMenuVisible` as a condition for rendering the `<CreateMenu>` Component? Something like this: 

```
{this.state.isMenuVisible ?  <CreateMenu... /> : null}
```

**Review by Julesssss — DISMISSED** [review]
Looking good 👍 

Apart from a few minor comments, I think the main change necessary is an improvement to the hack that @thienlnam pointed out.

File: src/components/CreateMenu.js
Comment (Maftalion) [reply src/components/CreateMenu.js:76]: yup, exactly. I originally had it defined outside but had to pull it in for the attachment logic

File: src/components/CreateMenu.js
Comment (Maftalion) [reply src/components/CreateMenu.js:16]: I don't believe we can do this unless  `MENU_ITEM_KEYS` is broken out of the `CONST` object with its own named export.

**Comment by thienlnam** [comment]
Tried to test the PR but it doesn't seem like the Add Attachment button is working for me at all even with the timeout

**Comment by Maftalion** [comment]
> Tried to test the PR but it doesn't seem like the Add Attachment button is working for me at all even with the timeout

hmm looks the change to conditionally render the component with `{this.state.isMenuVisible ? (` causes it to not work. going to investigate further.

**Comment by Maftalion** [comment]
@Julesssss looks like conditionally rendering using `this.state.isMenuVisible` causes the component to unmount before it can call `onModalHide`.
I can fix this by adding:
```
    componentWillUnmount() {
        this.onModalHide();
    }
```
let me know if this is fine, otherwise I can revert the conditional rendering.

I haven't been able to get the modals to work without using setTimeout, what's strange is `onModalHide` is supposed to be the solution to opening a modal once another closes so not sure why it's not working.

**Comment by Julesssss** [comment]
@Maftalion ah, lets revert to the previous solution then.

I'm also not sure why the callback doesn't work without setTimeout, it might be worth sharing the issue in the Contributor chat  slack channel to see if anyone else has an idea.

File: src/pages/home/report/ReportActionCompose.js
Comment (Julesssss) [reply src/pages/home/report/ReportActionCompose.js:194]: Ah good point, I forgot to mention but we have a new icon for this.

**Comment by Julesssss** [comment]
Hey @Expensify/design, would you mind sharing the new plus icon asset that will replace the paperclip icon. Shown in the following screenshot: 

Chat (4)

**Comment by Julesssss** [comment]
Hmm, the `setTimeout` issue is a tricky one.

I tried looking for a specific callback that was closing  the AttachmentPicker with no luck — the ReportActionCompose Component props aren’t changing (I thought this might be the issue) 

finally, the issue might lie within the [native AttachmentPicker](https://github.com/Expensify/Expensify.cash/blob/master/src/components/AttachmentPicker/index.js) Component as this issue doesn't exist on web, but I couldn't find anything here either 😞 

I’m a bit confused by the nesting of <CreateMenu> within <AttachmentPicker>  and wonder if this is necessary/could be made simpler?

**Comment by Maftalion** [comment]
`I’m a bit confused by the reverse nesting of within and wonder if this is necessary/could be made simpler? This isn't in scope for this PR though` - yeah, definitely makes it tricky to debug as well

Regarding the plus asset, I noticed we already have it so I can just swap it in. Do you also want the border gone like your screenshot above or is this fine?

Screen Shot 2021-02-27 at 12 25 03 AM

**Comment by shawnborton** [comment]
@Maftalion your screenshot looks perfect, let's go with that.

**Comment by Julesssss** [comment]
Hey @Maftalion. just wanted to check in and provide some guidance for moving forward.

As we'll need to resolve the issue without the `setTimeout` workaround, the next step will be to locate the cause (bug in Expensify.cash OR bug in external library). If this is an internal bug we should aim to fix it as part of the PR, but in the case where this is an external issue we could leave the PR on hold and mark the UpWork complete. 

We would then aim to resolve the issue within the offending library and if this was of interest to you we would treat this as a separate issue and provide an additional payment. Once the library issue is fixed, this PR can be taken off hold and merged.

**Comment by Maftalion** [comment]
Hey @Julesssss, I've been unable to get it to work without the setTimeout.

I've tried looking for re-render issues, moving things around and calling the Api directly instead of `openPicker` but all attempts were unsuccessful. 

I then looked into the `react-native-image-picker` and there seems to be major breaking changes between the current version and the latest so the whole actionsheet would need to be redone in order to test the latest version. But i also tested swapping it out with RN's built in `ActionSheetIOS` and that led to the same behavior.

Which meant the issue was probably with the modal library `react-native-modal`. There seems to be a lot of similar issues as this: https://github.com/react-native-modal/react-native-modal/issues?q=is%3Aissue+onmodalhide and a few lead to the same solution of just using `setTimeout`.

**Comment by Julesssss** [comment]
Hey @Maftalion, apologies for the delayed response. Thanks for looking into the modal library and potential solutions. 

Lets leave this PR on hold while I try to get the `react-native-modal` library issue created. I'll update this PR once I have an update.

**Comment by Julesssss** [comment]
Hi @Maftalion. I'm sorry for keeping you waiting for so long! 

To move forward, we'll merge this PR with the `setTimeout` workaround.

Would you mind fixing the conflicts with master to finish up here? I'll make sure to add a bonus to the payout as this is due to my delay.

**Comment by Maftalion** [comment]
@Julesssss no worries! fixed the merge conflicts

File: src/pages/home/report/ReportActionCompose.js
Comment (Julesssss) [inline src/pages/home/report/ReportActionCompose.js:245]: Ah, one final change... would you mind reducing this to 10. Just to avoid as much unnecessary delay as possible.

File: src/pages/home/report/ReportActionCompose.js
Comment (Julesssss) [reply src/pages/home/report/ReportActionCompose.js:227]: We decided to merge with this workaround. Another issue has been opened to improve the workaround.

**Review by Julesssss — APPROVED** [review]

**Comment by Julesssss** [comment]
Okay, we're almost there. 

I'm going to branch from here to prepare a PR locking the feature behind the new internal IOU Beta flag, once this is completed I will merge this issue and quickly follow it up by merging the beta flag PR. 

Should be first thing Monday.

**Comment by isagoico** [comment]
Hello! @Julesssss @Maftalion 
We were about to do the QA for this PR and came across this other PR https://github.com/Expensify/Expensify.cash/pull/1548. On the latter the options that we need to test for this one are actually expected to be hidden right? 
This would block us from being to execute the steps for this PR.

**Comment by Julesssss** [comment]
Hi @isagoico. 

Yeah, good point. Let's skip testing for this one in that case. I'll add the test steps to the PR which re-enables the new options.

**Pull request merged**
